iaas/digitalocean/digitalocean.go

```diff
@@ -0,0 +1,397 @@
+package digitalocean
+
+import (
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/pem"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"crypto/rand"
+
+	"github.com/digitalocean/godo"
+	"github.com/nuveo/gofn/iaas"
+	uuid "github.com/satori/go.uuid"
+	"golang.org/x/crypto/ssh"
+	"golang.org/x/oauth2"
+)
+
+var (
+	keysDir        = "./.gofn/keys"
+	privateKeyName = "id_rsa"
+	publicKeyName  = "id_rsa.pub"
+	sshPort        = ":22"
+)
+
+// Digitalocean difinition
+type Digitalocean struct {
+	client *godo.Client
+}
+
+// Auth in Digitalocean API
+func (do *Digitalocean) Auth() (err error) {
+	apiURL := os.Getenv("DIGITALOCEAN_API_URL")
+	key := os.Getenv("DIGITALOCEAN_API_KEY")
+	if key == "" {
+		err = errors.New("You must provide a Digital Ocean API Key")
+		return
+	}
+	tokenSource := oauth2.StaticTokenSource(&oauth2.Token{
+		AccessToken: key,
+	})
+	oauthClient := oauth2.NewClient(oauth2.NoContext, tokenSource)
+	do.client = godo.NewClient(oauthClient)
+	if apiURL != "" {
+		do.client.BaseURL, err = url.Parse(apiURL)
+		if err != nil {
+			return
+		}
+	}
+	return
+}
+
+// CreateMachine on digitalocean
+func (do *Digitalocean) CreateMachine() (machine *iaas.Machine, err error) {
+	err = do.Auth()
+	if err != nil {
+		return
+	}
+	snapshots, _, err := do.client.Snapshots.List(nil)
+	if err != nil {
+		return
+	}
+
+	snapshot := godo.Snapshot{}
+	for _, s := range snapshots {
+		if s.Name == "GOFN" {
+			snapshot = s
+			break
+		}
+	}
+	image := godo.DropletCreateImage{
+		Slug: "debian-8-x64",
+	}
+	if snapshot.Name != "" {
+		id, _ := strconv.Atoi(snapshot.ID)
+		image = godo.DropletCreateImage{
+			ID: id,
+		}
+	}
+
+	sshKey, err := do.getSSHKeyForDroplet()
+	if err != nil {
+		return
+	}
+	createRequest := &godo.DropletCreateRequest{
+		Name:   fmt.Sprintf("gofn-%s", uuid.NewV4().String()),
+		Region: "nyc1",
+		Size:   "512mb",
```

Comment:
Reviewed in https://github.com/nuveo/gofn/commit/ca67a0ef3751b9199d3e8cb1abccdab63dcff877